# Mini-Project                   Apriori Algorithm

ref https://stackabuse.com/association-rule-mining-via-apriori-algorithm-in-python/

In [1]:
                                                                        #importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from apyori import apriori                                              #apyori library

In [2]:
store_data = pd.read_excel('train.xlsx')                                #creating data frame of training data

In [3]:
store_data.head()

EmployeeNumber  Age     BusinessTravel              Department  Education  \
0               1   41      Travel_Rarely                   Sales          2   
1               2   49  Travel_Frequently  Research & Development          1   
2               4   37      Travel_Rarely  Research & Development          2   
3               5   33  Travel_Frequently  Research & Development          4   
4               7   27      Travel_Rarely  Research & Development          1   

  EducationField  Gender  JobLevel                JobRole  JobSatisfaction  \
0  Life Sciences  Female         2        Sales Executive                4   
1  Life Sciences    Male         2     Research Scientist                2   
2          Other    Male         1  Laboratory Technician                3   
3  Life Sciences  Female         1     Research Scientist                3   
4        Medical    Male         1  Laboratory Technician                2   

  MaritalStatus  MonthlyIncome  NumCompaniesWorked OverTime  \
0        Single           5993                   8      Yes   
1       Married           5130                   1       No   
2        Single           2090                   6      Yes   
3       Married           2909                   1      Yes   
4       Married           3468                   9       No   

   PercentSalaryHike  PerformanceRating  TotalWorkingYears  YearsAtCompany  \
0                 11                  3                  8               6   
1                 23                  4                 10              10   
2                 15                  3                  7               0   
3                 11                  3                  8               8   
4                 12                  3                  6               2   

   YearsSinceLastPromotion Attrition  
0                        0       Yes  
1                        1        No  
2                        0       Yes  
3                        3        No  
4                        2        No

In [4]:
store_data = pd.read_excel('train.xlsx', header=None)                     #removing header

In [5]:
store_data.head()

0    1                  2                       3          4   \
0  EmployeeNumber  Age     BusinessTravel              Department  Education   
1               1   41      Travel_Rarely                   Sales          2   
2               2   49  Travel_Frequently  Research & Development          1   
3               4   37      Travel_Rarely  Research & Development          2   
4               5   33  Travel_Frequently  Research & Development          4   

               5       6         7                      8                9   \
0  EducationField  Gender  JobLevel                JobRole  JobSatisfaction   
1   Life Sciences  Female         2        Sales Executive                4   
2   Life Sciences    Male         2     Research Scientist                2   
3           Other    Male         1  Laboratory Technician                3   
4   Life Sciences  Female         1     Research Scientist                3   

              10             11                  12        13  \
0  MaritalStatus  MonthlyIncome  NumCompaniesWorked  OverTime   
1         Single           5993                   8       Yes   
2        Married           5130                   1        No   
3         Single           2090                   6       Yes   
4        Married           2909                   1       Yes   

                  14                 15                 16              17  \
0  PercentSalaryHike  PerformanceRating  TotalWorkingYears  YearsAtCompany   
1                 11                  3                  8               6   
2                 23                  4                 10              10   
3                 15                  3                  7               0   
4                 11                  3                  8               8   

                        18         19  
0  YearsSinceLastPromotion  Attrition  
1                        0        Yes  
2                        1         No  
3                        0        Yes  
4                        3         No

In [20]:
records = []                                                              #preprocessing of data into List
for i in range(1,15):
    records.append([str(store_data.values[i,j]) for j in [1,3,9,10,-1]])

In [21]:
records

[['41', 'Sales', '4', 'Single', 'Yes'],
 ['49', 'Research & Development', '2', 'Married', 'No'],
 ['37', 'Research & Development', '3', 'Single', 'Yes'],
 ['33', 'Research & Development', '3', 'Married', 'No'],
 ['27', 'Research & Development', '2', 'Married', 'No'],
 ['32', 'Research & Development', '4', 'Single', 'No'],
 ['59', 'Research & Development', '1', 'Married', 'No'],
 ['30', 'Research & Development', '3', 'Divorced', 'No'],
 ['38', 'Research & Development', '3', 'Single', 'No'],
 ['36', 'Research & Development', '3', 'Married', 'No'],
 ['35', 'Research & Development', '2', 'Married', 'No'],
 ['29', 'Research & Development', '3', 'Single', 'No'],
 ['31', 'Research & Development', '3', 'Divorced', 'No'],
 ['34', 'Research & Development', '4', 'Divorced', 'No']]

['Age', 'Department', 'JobSatisfaction', 'MaritalStatus', 'Attrition']

In [22]:
                                                                          #genearating association rules
association_rules = apriori(records, min_support=0.07, min_confidence=0.7, min_lift=3, min_length=3)
association_results = list(association_rules)

In [23]:
association_results

[RelationRecord(items=frozenset({'1', '59'}), support=0.07142857142857142, ordered_statistics=[OrderedStatistic(items_base=frozenset({'1'}), items_add=frozenset({'59'}), confidence=1.0, lift=14.0), OrderedStatistic(items_base=frozenset({'59'}), items_add=frozenset({'1'}), confidence=1.0, lift=14.0)]),
 RelationRecord(items=frozenset({'27', '2'}), support=0.07142857142857142, ordered_statistics=[OrderedStatistic(items_base=frozenset({'27'}), items_add=frozenset({'2'}), confidence=1.0, lift=4.666666666666667)]),
 RelationRecord(items=frozenset({'35', '2'}), support=0.07142857142857142, ordered_statistics=[OrderedStatistic(items_base=frozenset({'35'}), items_add=frozenset({'2'}), confidence=1.0, lift=4.666666666666667)]),
 RelationRecord(items=frozenset({'49', '2'}), support=0.07142857142857142, ordered_statistics=[OrderedStatistic(items_base=frozenset({'49'}), items_add=frozenset({'2'}), confidence=1.0, lift=4.666666666666667)]),
 RelationRecord(items=frozenset({'Divorced', '30'}), suppo

In [24]:
print(association_results[0])                                             #printing first association rule

RelationRecord(items=frozenset({'1', '59'}), support=0.07142857142857142, ordered_statistics=[OrderedStatistic(items_base=frozenset({'1'}), items_add=frozenset({'59'}), confidence=1.0, lift=14.0), OrderedStatistic(items_base=frozenset({'59'}), items_add=frozenset({'1'}), confidence=1.0, lift=14.0)])


In [25]:
for item in association_results:                                          #printing association rules
    pair = item[0] 
    items = [x for x in pair]
 
    if(len(items)==2):
        print (items[0] + '->' + items[1])
    if(len(items)==3):
        print (items[0] + '->' + items[1] + '->' + items[2])
    if(len(items)==4):
        print (items[0] + '->' + items[1] + '->' + items[2] + '->' + items[3])
    if(len(items)>4):
        print (items[0] + '->' + items[1] + '->' + items[2] + '->' + items[3] + '->' + items[4])
        
    print("Support: " + str(item[1]))
    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("=====================================")

1->59
Support: 0.07142857142857142
Confidence: 1.0
Lift: 14.0
27->2
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
35->2
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
49->2
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
Divorced->30
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
Divorced->31
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
32->4
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
4->34
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
Divorced->34
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
37->Yes
Support: 0.07142857142857142
Confidence: 1.0
Lift: 7.0
4->41
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
4->Sales
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667
Sales->41
Support: 0.07142857142857142
Confidence: 1.0
Lift: 14.0
Yes->41
Support: 0.0714285714285714

Example of Association Rules:

Divorced->No->30->Research & Development
Support: 0.07142857142857142
Confidence: 1.0
Lift: 4.666666666666667

=====================================

37->Yes->Single->Research & Development
Support: 0.07142857142857142
Confidence: 1.0
Lift: 14.0